src/MechaSwap.sol

```diff
@@ -0,0 +1,140 @@
+    /// @dev reserve token X
+    IERC20 public immutable tokenX;
+    /// @dev reserve token Y
+    IERC20 public immutable tokenY;
+
+
+    /// @notice emitted when a user contributes liquidity to the pool
+    /// @param user address of user that is providing liquidity to the pool
+    /// @param xIn amount of token X that is being provided as liquidity
+    /// @param yIn amount of token Y that is being provided as liquidity 
+    /// @param zOut amount of LP tokens that are being minted to the user
+    event LiquidityProvided(address indexed user, uint256 xIn, uint256 yIn, uint256 zOut);
+
+    /// @notice emitted when a user removes liquidity from the pool
+    /// @param user address of user that is removing liquidity from the pool
+    /// @param xOut amount of token X that is being removed from liquidity
+    /// @param yOut amount of token Y that is being removed from liquidity
+    /// @param zIn amount of LP tokens that are being burned from the user
+    event LiquidityRemoved(address indexed user, uint256 xOut, uint256 yOut, uint256 zIn);
+
+    /// @notice emitted when a usere swaps tokens within the pool
+    /// @param user address of user that is swapping tokens
+    /// @param xIn amount of token X that is being swapped into the pool
+    /// @param yIn amount of token Y that is being swapped into the pool
+    /// @param xOut amount of token X that is being swapped out of the pool
+    /// @param yOut amount of token Y that is being swapped out of the pool
+    event Swap(address indexed user, uint256 xIn, uint256 yIn, uint256 xOut, uint256 yOut);
+
+    constructor(IERC20 tokenX_, IERC20 tokenY_) ERC20("MechaSwapLP", "MSLP", 18) {
+        tokenX = tokenX_;
+        tokenY = tokenY_;
+    }
+
+    /// @notice used to intiate a liquidity pool, can only be done once
+    /// @param xIn the amount of token X that'll be added to the pool 
+    /// @param yIn the amount of token Y that'll be added to the pool
+    /// @return zOut the amount of LP tokens that'll be minted to the user
+    function init(uint256 xIn, uint256 yIn) external returns(uint256 zOut){
+        require(_totalSupply == 0, "MechaSwap:Pool already initiated");
+        require(xIn > 0 && yIn > 0, "MechaSwap: Can't provide 0 tokens");
+        x_0 += xIn;
+        y_0 += yIn;
+        zOut = xIn * yIn;
+        tokenX.safeTransferFrom(msg.sender, address(this), xIn);
+        tokenY.safeTransferFrom(msg.sender, address(this), yIn);
+        _mint(msg.sender, zOut);
+        emit LiquidityProvided(msg.sender, xIn, yIn, zOut);
+        }
+
+    /// @notice used to add liquidity to an existing pool
+    /// @param xIn the amount of token X that'll be added to the pool
+    /// @return zOut the amount of LP tokens that'll be minted to the user
+    function addLiquidity(uint256 xIn) external returns(uint256 zOut){
+        require(_totalSupply > 0, "MechaSwap:Pool not intiated");
+        uint256 _x_0 = x_0;
+        uint256 _y_0 = y_0;
+        if(_x_0 > _y_0){
```

Comment:
Previously, if X reserves were more than Y reserves, and you tried to add liquidity with xIn as 1, it would result in an underflow error. This prevents that but not sure if it matters? 
Also added the _x_0 and _y_0 to prevent multiple SLOADs

Reply:
I'm not sure it matters, either. Could you describe the scenario with more detail?

Reply:
I've thought about it more and I don't think it does cause it's not a pool were the tokens are intended to have 1:1 balance in reserves. 
As I've implemented it, it only works for when token X has a larger balance in reserves than token Y which isn't guaranteed.
During testing at init token X had 1000 times the balance in the pool as token Y. So if you tried to add 999 wei or lower of token X as liquidity it would cause an underflow. Since this causes the transaction to revert anyways I don't think it's a problem and effort would be better put into improving the UX when adding liquidity